### PR TITLE
[Rust Shim] added support for rust .crs files

### PR DESCRIPTION
**Pull request by deadjakk**
I added support for a rust shim that I wrote ([associated crates.io link](https://crates.io/crates/metasploit-shim)), to bridge.rb. 
This will allow msf to recognize '.crs' files and execute them using the `cargo script` ( subcommand [associated crates.io link](https://crates.io/crates/cargo-script)) which will build and execute modules written in rust such as the example pushed [here](https://github.com/deadjakk/metasploit-rust-shim/blob/main/example-module/src/example-module.crs).

### Pre-requisite installation 
   - [ ]  Install cargo, the rust toolchain
   - [ ]  Install the `script` subcommand for cargo (you can install this using `cargo install cargo-script` once you have cargo installed.
### Run the module
   - [ ]  Replace the bridge.rb file located in /lib/msf/core/modules/external/bridge.rb with the file provided in this repo.
   - [ ]  Place the example .crs file in ~/.msf4/modules/exploits/, or wherever you'd like to store the module.
   - [ ]  Start `msfconsole`
   - [ ] Use the module as usual

The screenshot below shows the full code of an example module and the same module being run from msfconsole. (sorry about the weird dimensions).

rust-shim-working

tagging @wvu-r7 per request

### CODE REVIEW

File: lib/msf/core/modules/external/bridge.rb [hunk 1]
```diff
@@ -238,12 +238,33 @@ def handle_exception(error)
...
+
+  def initialize(module_path, framework: nil)
+    super
+    self.cmd = ['cargo', 'script', self.path]
+  end
+
+  def handle_exception(error)
+    case error
+    when Errno::ENOENT
+      LoadError.new('Failed to execute external Go module. Please ensure you have Go installed on your environment.')
```
Comment (adfoster-r7) [inline lib/msf/core/modules/external/bridge.rb:254]: Looks like this line needs updated 👀

**Comment by adfoster-r7** [comment]
Looks good to me after a quick glance. I'll need to look at this in more detail later, but I also seem to remember there being some extra logic in msf which relies on shebangs being present for to handle external modules correctly:

https://github.com/rapid7/metasploit-framework/blob/3c1a6a480cf01921c81756ed40da24f379e6a2a8/lib/msf/core/modules/loader/executable.rb#L14-L17

https://github.com/rapid7/metasploit-framework/blob/3c1a6a480cf01921c81756ed40da24f379e6a2a8/lib/msf/core/modules/loader/base.rb#L468-L473

Which I believe the crs files don't have

**Comment by deadjakk** [comment]
Explanation regarding the location of the shim library.: the reason I placed the rust code in an external [location](https://crates.io/crates/metasploit-shim), rather than add a directory under .../external/rust is because cargo-script doesn't support relative path dependencies. Crates.io is permanent, no changes can be made now or in the future for metasploit-shim/0.1.0.

**Comment by deadjakk** [comment]
> Looks good to me after a quick glance. I'll need to look at this in more detail later, but I also seem to remember there being some extra logic in msf which relies on shebangs being present for it to handle external modules correctly:
> 
> https://github.com/rapid7/metasploit-framework/blob/3c1a6a480cf01921c81756ed40da24f379e6a2a8/lib/msf/core/modules/loader/executable.rb#L14-L17
> 
> https://github.com/rapid7/metasploit-framework/blob/3c1a6a480cf01921c81756ed40da24f379e6a2a8/lib/msf/core/modules/loader/base.rb#L468-L473
> 
> Which I believe the crs files don't have
I was also able to get the .crs file to run using 
`#!/usr/bin/env -S cargo script `
but from what I could tell, msf will just launch it using the `cargo script` line in the provided bridge

**Comment by adfoster-r7** [comment]
I think when msfconsole loads initially, metasploit will try detecting new modules to populate the search cache with. In your example screenshot, the module is loaded directly with `use` - which I think goes down a slightly different codepath which doesn't rely on the `script_path?` functionality

I could be wrong though, and still need to verify those semantics 🔍

**Comment by deadjakk** [comment]
> I think when msfconsole loads initially, metasploit will try detecting new modules to populate the search cache with. In your example screenshot, the module is loaded directly with `use` - which I think goes down a slightly different codepath which doesn't rely on the `script_path?` functionality
> 
> I could be wrong though, and still need to verify those semantics mag

My understand was that when msfconsole was run, it sees that the .crs is an executable file and then compiles/executes it and then sends a 'method':'describe' via stdin (JSON RPC). the executed module then responds via stdout (RPC JSON) with the metadata with which the search cache is populated.

use -> run appears to perform he same steps but sends a 'method':'run' in the stdin JSON RPC, prompting the module to run a separate callback function.

**Comment by jmartin-r7** [comment]
> Explanation regarding the location of the shim library.: the reason I placed the rust code in an external [location](https://crates.io/crates/metasploit-shim), rather than add a directory under .../external/rust is because cargo-script doesn't support relative path dependencies. Crates.io is permanent, no changes can be made now or in the future for metasploit-shim/0.1.0.

While this is understandable, the limitation may need to be overcome in some other manner, one idea might be to include the source in a non relative way.  One of the concerns here is that a 0.1.1 with some malicious component could possibly be created and a module offered that is set to utilize this version could cause a supply chain issue. While this is not a highly likely scenario, even your example modules seems to be [constructed in a way](https://github.com/deadjakk/metasploit-rust/blob/9a00b98b85d8fb60f43c1f4738dae7761040c5c0/example-module/Cargo.toml#L9-L10) that would be susceptible to this since it does not lock the expected dependency version.

I would also ask that you consider adjusting the licensing for [metasploit-shim](https://github.com/deadjakk/metasploit-rust) to be a bit more permissive.  The project prefers BSD3, MIT, or Apache whenever possible and all the dependencies in `metasploit-shim` are dual licensed as MIT & Apache.

**Comment by adfoster-r7** [comment]
I know this is still in the initial stages of implementing metasploit's json api in rust, but out of interest - are there any Rust modules that would also be good candidates for integrating with Metasploit? 👀

**Comment by deadjakk** [comment]
> > Explanation regarding the location of the shim library.: the reason I placed the rust code in an external [location](https://crates.io/crates/metasploit-shim), rather than add a directory under .../external/rust is because cargo-script doesn't support relative path dependencies. Crates.io is permanent, no changes can be made now or in the future for metasploit-shim/0.1.0.
> 
> While this is understandable, the limitation may need to be overcome in some other manner, one idea might be to include the source in a non relative way. One of the concerns here is that a 0.1.1 with some malicious component could possibly be created and a module offered that is set to utilize this version could cause a supply chain issue. While this is not a highly likely scenario, even your example modules seems to be [constructed in a way](https://github.com/deadjakk/metasploit-rust/blob/9a00b98b85d8fb60f43c1f4738dae7761040c5c0/example-module/Cargo.toml#L9-L10) that would be susceptible to this since it does not lock the expected dependency version.
> 
> I would also ask that you consider adjusting the licensing for [metasploit-shim](https://github.com/deadjakk/metasploit-rust) to be a bit more permissive. The project prefers BSD3, MIT, or Apache whenever possible and all the dependencies in `metasploit-shim` are dual licensed as MIT & Apache.

I updated the crates library (and associated repo) to contain an MIT license instead of GPL in metasploit_shim = "0.1.3"

As for the concern regarding the security of the version being used, metasploit_shim = "0.1.3", instead of metasploit_shim = "*", would ensure that only a specific version is used.

**Comment by adfoster-r7** [comment]
@deadjakk Were there any plans for modules beyond the helloworld example that would be good to develop? 👀

**Comment by wvu-r7** [comment]
Hello World exploit.

**Comment by wvu-r7** [comment]
`HELLOWORLD` exploit.

**Comment by deadjakk** [comment]
> @deadjakk Were there any plans for modules beyond the helloworld example that would be good to develop? eyes

Yes I am working on one currently

**Comment by sempervictus** [comment]
In terms of the absolution location bit, we might be able to hack around that with submodules to create relative paths. There have been issues opened for this functionality multiple times, at least one currently up right now. For all the things Rust does intelligently it does occasionally have a tendency to trip over its own stiff upper lip.

**Comment by jmartin-r7** [comment]
We're going to attic this for now; I don't think we're in a position to accept a Rust external module integration just yet, as there isn't enough benefit to warrant the overhead. We could potentially accept a future pull request to allow the existing plugin system to introduce changes to the external module loader though, which could be a stepping stone for the continued development of Rust modules that live outside of Metasploit's codebase.

In the current state the proposed module is completely external and does not seem friendly to get up and running. If this were to be enhanced or reworked to provide a plugin enhancing the module loader it might be usable, however managing modules from the external source would be up to the maintainer of the plugin.  The current proposed module is also missing significant metadata though and if not in the framework repo that would not be a quality bar for the external repository.

One of the concerns here is for me is the workflow and the integration points are just not useful yet. The modules would need to fit somewhere in frameworks pathing structures, and modules written in Rust need to get some value from the language as well as get some value from the framework beyond inclusion.

The module offered is not gaining any value from metasploit framework. The suggested module can be implemented in `Ruby` taking advantage of the framework command execution libraries and does not gain an advantage from using `Rust` other than possibly leveraging an author’s exiting comfort in the language.

Consider [`exploit/linux/http/pulse_secure_cmd_exec`](https://github.com/rapid7/metasploit-framework/blob/8d808d11c0adc61d8ed3569ae41a17f5e678e998/modules/exploits/linux/http/pulse_secure_cmd_exec.rb) seems to have similar characteristics to the CVE of the proposed module.  However this module gains capabilities from metasploit framework by using the CmdStager mixin library to expand and operationalize it to be able to deliver not just commands but to stage an interactive payload session. I am not sure this is possible with your vulnerable application but it is an example of the module gaining something from metasploit framework.

**Pull request closed without merging**